experimental/memory.go

```diff
@@ -6,18 +6,34 @@ import (
 	"github.com/tetratelabs/wazero/internal/expctxkeys"
 )
 
-// MemoryAllocator is a memory allocation hook which is invoked
-// to create a new MemoryBuffer, with the given specification:
-// min is the initial and minimum length (in bytes) of the backing []byte,
-// cap a suggested initial capacity, and max the maximum length
-// that will ever be requested.
-type MemoryAllocator func(min, cap, max uint64) MemoryBuffer
+// MemoryAllocator is a memory allocation hook,
+// invoked to create a LinearMemory.
+type MemoryAllocator interface {
+	// Allocate should create a new LinearMemory with the given specification:
+	// min is the initial and minimum length (in bytes) of the backing []byte,
+	// cap a suggested initial capacity, and max the maximum length
+	// that will ever be requested.
+	//
+	// Notes:
+	//   - To back a shared memory, the address of the backing []byte cannot
+	//     change. This is checked at runtime. Implementations should document
+	//     if the returned LinearMemory meets this requirement.
+	Allocate(min, cap, max uint64) LinearMemory
+}
+
+// MemoryAllocatorFunc is a convenience for defining inlining a MemoryAllocator.
+type MemoryAllocatorFunc func(min, cap, max uint64) LinearMemory
+
+// Allocate implements MemoryAllocator.Allocate.
+func (f MemoryAllocatorFunc) Allocate(min, cap, max uint64) LinearMemory {
+	return f(min, cap, max)
+}
 
-// MemoryBuffer is a memory buffer that backs a Wasm memory.
-type MemoryBuffer interface {
-	// Buffer returns the backing []byte for the memory buffer.
+// LinearMemory is an expandable []byte that backs a Wasm linear memory.
+type LinearMemory interface {
+	// Buffer returns the backing []byte for the linear memory.
```

Comment:
What were your thoughts on dropping this method in favor of Grow(0)?

Reply:
I don't mind either way. Smaller interfaces are good. OTOH we're not growing the memory.

I also thought grow might not return the buffer. Or… endless possibilities.

Reply:
This is, maybe, a little less intuitive. `Grow(size)` grows the memory to `size`.
So it's not `Grow(0)` it's `Grow(min)`.

And then it probably makes little sense to have `Allocate(min, cap, max uint64) LinearMemory`, rather than just `Allocate(cap, max uint64) LinearMemory`.

Reply:
Ah OK, `Grow` is maybe non-intuitive then because Go developers will probably be familiar with [`bytes.(*Buffer).Grow`](https://pkg.go.dev/bytes#Buffer.Grow) which takes the number of bytes to make space for, not the total size. Same for the [`memory.grow`](https://developer.mozilla.org/en-US/docs/WebAssembly/Reference/Memory/Grow) instruction in WebAssembly.

How about renaming the method to `Resize`, or making it work like `bytes.(*Buffer).Grow`?